libbeat/common/mapstr.go

```diff
@@ -305,25 +305,38 @@ func MergeFields(ms, fields MapStr, underRoot bool) error {
 // exist then it will be created. If the tags field exists and is not a []string
 // then an error will be returned. It does not deduplicate the list of tags.
 func AddTags(ms MapStr, tags []string) error {
+	return AddTagsWithKey(ms, TagsKey, tags)
+}
+
+// AddTagsWithKey appends a tag to the key field of ms. If the field does not
+// exist then it will be created. If the field exists and is not a []string
+// then an error will be returned. It does not deduplicate the list.
+func AddTagsWithKey(ms MapStr, key string, tags []string) error {
 	if ms == nil || len(tags) == 0 {
 		return nil
 	}
-	eventTags, exists := ms[TagsKey]
-	if !exists {
-		ms[TagsKey] = tags
+
+	k, subMap, oldTags, present, err := mapFind(key, ms, true)
```

Comment:
I wodner if this has a perfomance impact on the old `AddTags` method.

Reply:
No. Keys without dots can be processed in the first loop on a fast path in `mapFind`. If the key exists as is, it's simply returned at the beginning.

```golang
// Fast path, key is present as is.
if v, exists := data[key]; exists {
    return key, data, v, true, nil
}
```

If not, the only additional cost is one `IndexRune` which looks for a dot in the key. But currently old tags don't have dots in their keys. Thus, the function returns at the second possible point without doing anything "expensive" e.g. creating additional submaps.

```golang
idx := strings.IndexRune(key, '.')
if idx < 0 {
    return key, data, nil, false, nil
}
```